### ISSUE TITLE
Add dim to view_as_real and view_as_complex

### ISSUE THREAD

**Issue by CerebralSeed** [pytorch/pytorch#71066]
### 🚀 The feature, motivation and pitch

I'm working on developing convolution operations for complex numbers and it would be helpful if view_as_real and view_as_complex had a dim which allowed you to specify where the split would occur. 

For example, say I want to create the following class:

```
class Conv2dComp(nn.Module):
    def __init__(self, in_channels, out_channels, kernel_size, stride=1, padding=0, dilation=1, groups=1, bias=True, padding_mode='zeros'):
        super(Conv2dComp, self).__init__()
        self.conv=nn.Conv2d(in_channels*2, out_channels, kernel_size, stride, padding, dilation, groups*2, bias, padding_mode)
    def forward(self, z):
        return torch.view_as_complex(self.conv(torch.view_as_real(z)))

testconv=Conv2dComp(3, 10, (5, 5))
A=torch.rand(16, 3, 28, 28, dtype=torch.cfloat)
print(testconv(A))
```

This actually won't work because torch.view_as_real(z).size() yields ([16, 3, 28, 28, 2]).

But what I want to be able to do is specify a dim to split the real and imaginary into reals. Say torch.view_as_real(z, dim=1) which then would be expected to yield size of ([16, 6, 28, 28]). In other words, just a way to specify where those get split. Of course, the default could be set to dim=-1 so it does exactly what it does now, if left unspecified. 

And conversely for view_as_complex - in order to specify the dim where the real vs. imaginary split should occur. 

Of course, I understand you'll have to define which values in a given dimension should become real and which should become imaginary. Currently, the view_as_complex operation requires a final dim of 2. So how would one deal with a tensor where the dim is greater than 2? Such as ([16, 6, 28, 28]) where we want to target dim=1?

In that case, depending on how feasible, there could be specified an option such as "split='pairs'" or "split='half'". This would be such that:

```
B=torch.rand(3, 6)

tensor([[0.5981, 0.7810, 0.4593, 0.7290, 0.9865, 0.4719],
        [0.7933, 0.0734, 0.0311, 0.6109, 0.8687, 0.4748],
        [0.7676, 0.0304, 0.2736, 0.4384, 0.9076, 0.0597]])

print(view_as_real(B, dim=1, split='pairs'))

tensor([[0.5981+0.7810j, 0.4593+0.7290j, 0.9865+0.4719j],
        [0.7933+0.0734j, 0.0311+0.6109j, 0.8687+0.4748j],
        [0.7676+0.0304j, 0.2736+0.4384j, 0.9076+0.0597j]])

print(view_as_real(B, dim=1, split='half'))

tensor([[0.5981+0.7290j, 0.7810+0.9865j, 0.4593+0.4719j],
        ...
         ])
```

cc @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved

**Comment by CerebralSeed** [comment]
> You should be able to do this with the generic `view`, `reshape`, `split`, etc. APIs right?

It's totally doable. 

For example:

```
A=torch.rand(1, 3, 5, 5, dtype=torch.cfloat)

B=torch.view_as_real(A).permute(0,1,4,2,3).reshape(1,6,5,5)

print(A)
print(B)
```

Just a bit messy and probably not the most efficient way, timewise. Hence the feature request.

**Comment by anjali411** [comment]
Hi @CerebralSeed these functions were added as a way to help people transition to the complex tensors added to PyTorch in 1.6. Most people who have been using complex numbers from before that time, followed the convention of (..., 2) to represent complex numbers where the real and imag values are stored in the last dimension. As Greg mentioned, you can also get a complex tensor with an arbitrary dimension containing real and imag values using reshape and other view ops. To reiterate this point, I am gonna take the example you shared in the issue description.,
```
B=torch.rand(3, 6)
print(B)
print(torch.view_as_complex(B.reshape(3, 3, 2)))
``` 
out:
```
tensor([[0.2105, 0.6835, 0.7851, 0.8001, 0.2591, 0.3676],
        [0.3148, 0.0229, 0.7898, 0.1459, 0.4130, 0.9463],
        [0.0051, 0.4769, 0.5182, 0.1009, 0.4409, 0.0428]])
tensor([[0.2105+0.6835j, 0.7851+0.8001j, 0.2591+0.3676j],
        [0.3148+0.0229j, 0.7898+0.1459j, 0.4130+0.9463j],
        [0.0051+0.4769j, 0.5182+0.1009j, 0.4409+0.0428j]])
```

**Comment by Lezcano** [comment]
For what is worth, you mention @CerebralSeed  "Just a bit messy and probably not the most efficient way,". Note that both `permute` and `reshape` are views of the original tensor, so efficiency-wise they are as fast as they get.

**Comment by ezyang** [comment]
We can't do this for `view_as_complex`, because the complex dtype implies that the real and complex parts are stored adjacent to each other in memory. To allow a split would also require a whole other set of kernels for the split case.

That being said, this isn't *unreasonable* to want, it's just that the costs of actually building out this whole other memory layout outweight the marginal benefit of this case. Maybe some of the work we're doing might lower this cost or make it possible to do an out of tree implementation.
